Sprint 17

SergeySKRV opened the pull request.

SergeySKRV committed:
✨ feat: Added multilingual localization and improved tracker search

SergeySKRV committed:
✨ feat: Added tracker filtering functionality

SergeySKRV committed:
✨ feat: Add dark mode support and update screenshot tests

• Set up adaptive colors in assets (Any/Dark)
• Updated ViewControllers to use colors from assets
• Added screenshot tests for both light and dark themes
• Application now fully supports dark mode

SergeySKRV committed:
✨ feat: Add AppMetrica analytics to all application screens

SergeySKRV committed:
✨ feat: Implement Statistics screen and general code style improvements

IgorChumikov commented on Tracker/ViewControllers/ScheduleViewController.swift:
```diff
@@ -127,6 +153,16 @@ extension ScheduleViewController: UITableViewDataSource, UITableViewDelegate {
+                "item": "day_selected",
+                "day": day?.fullName ?? "Unknown"
+            ]
+            AppMetrica.reportEvent(name: "Screen Event", parameters: dayEvent)
```
Это работает, но есть минусы:

- легко сделать опечатку в ключах или значениях ("Filter" vs "Filters");
- при изменении названий придётся искать и исправлять строки по всему проекту;
- нет автодополнения в Xcode — нужно помнить точное написание;
- структура события дублируется в нескольких местах.

💡 Как лучше в реальных проектах
Обычно создают отдельный сервис для аналитики и используют enum для всех событий, экранов и элементов.
Это даёт:

- типобезопасность (ошибка сразу видна при компиляции, а не в аналитике);
- автодополнение и быструю навигацию;
- единый центр, где хранятся все события.

Пример улучшенной реализации:
```

import AppMetricaCore

final class AnalyticsService {
    static let shared = AnalyticsService()
    private init() {}

    func reportEvent(_ event: AnalyticsEvent) {
        var parameters: [String: Any] = [
            "event": event.type.rawValue,
            "screen": event.screen.rawValue
        ]
        
        if let item = event.item?.rawValue {
            parameters["item"] = item
        }
        
        AppMetrica.reportEvent(name: "user_action", parameters: parameters)
    }
}

// MARK: - Models

enum AnalyticsEventType: String {
    case open, close, click
}

enum AnalyticsScreen: String {
    case main = "Main"
    case statistics = "Statistics"
    case filter = "Filter"
}

enum AnalyticsItem: String {
    case addTrack = "add_track"
    case track = "track"
    case filter = "filter"
    case edit = "edit"
    case delete = "delete"
}

struct AnalyticsEvent {
    let type: AnalyticsEventType
    let screen: AnalyticsScreen
    let item: AnalyticsItem?
}

```

Использование:
```

AnalyticsService.shared.reportEvent(
    AnalyticsEvent(type: .open, screen: .filter, item: nil)
)

```

Твой способ рабочий и подойдёт для учебного проекта. Но в реальной разработке мы почти всегда используем сервис с enum, как в примере выше — это избавляет от опечаток, ускоряет работу и упрощает поддержку проекта

IgorChumikov commented on Tracker/AppDelegate.swift:
```diff
@@ -10,6 +11,11 @@ final class AppDelegate: UIResponder, UIApplicationDelegate {
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
         _ = coreDataStack.viewContext
+   
+        if let configuration = AppMetricaConfiguration(apiKey: "147a279e-5dcc-4356-a7bb-06b778f39284") {
```
🔐 API-ключ лучше вынести в константу — например, в Constants.swift, Secrets.swift или внутри AnalyticsService, чтобы:

избежать дублирования;
упростить управление конфигурацией;
облегчить смену ключа в будущем;
повысить читаемость и безопасность кода.

✅ Пример:
```

// Constants.swift
enum AnalyticsConstants {
    static let appMetricaAPIKey = "a2171b29-a56b-42eb-ae40-fba55935c332"
}

```
```
if let configuration = AppMetricaConfiguration(apiKey: AnalyticsConstants.appMetricaAPIKey) {
    AppMetrica.activate(with: configuration)
}
```

IgorChumikov commented on TrackerTests/TrackersViewControllerSnapshotTests.swift:
```diff
@@ -0,0 +1,103 @@
+import SnapshotTesting
+@testable import Tracker
+
+class TrackersViewControllerSnapshotTests: XCTestCase {
```
Я перепишу его с учётом: убрать дублирование, вынести vc в свойство, использовать setUp / tearDown и добавить // Given // When // Then.
```

import XCTest
import SnapshotTesting
@testable import Tracker

final class TrackersViewControllerSnapshotTests: XCTestCase {
    
    // MARK: - Properties
    
    private var viewController: TrackersViewController!
    private var mockViewModel: MockTrackersViewModel!
    
    // MARK: - Lifecycle
    
    override func setUp() {
        super.setUp()
        
        mockViewModel = MockTrackersViewModel()
        setupMockData(for: mockViewModel)
        viewController = TrackersViewController(viewModel: mockViewModel)
    }
    
    override func tearDown() {
        weak var weakVC = viewController
        viewController = nil
        mockViewModel = nil
        super.tearDown()
        
        XCTAssertNil(weakVC, "TrackersViewController should be deallocated. Potential memory leak.")
    }
    
    // MARK: - Tests
    
    func testTrackersViewControllerLightMode() {
        // When
        viewController.overrideUserInterfaceStyle = .light
        
        // Then
        assertSnapshot(of: viewController, as: .image(traits: .init(userInterfaceStyle: .light)), named: "light_mode")
    }
    
    func testTrackersViewControllerDarkMode() {
        // When
        viewController.overrideUserInterfaceStyle = .dark
        
        // Then
        assertSnapshot(of: viewController, as: .image(traits: .init(userInterfaceStyle: .dark)), named: "dark_mode")
    }
    
    // MARK: - Helpers
    
    private func setupMockData(for mockVM: MockTrackersViewModel) {
        guard let categoryId1 = UUID(uuidString: "00000000-0000-0000-0000-000000000001"),
              let categoryId2 = UUID(uuidString: "00000000-0000-0000-0000-000000000002") else {
            XCTFail("Failed to create UUIDs from strings")
            return
        }
        
        mockVM.stubVisibleCategories = [
            TrackerCategory(
                id: categoryId1,
                title: "Здоровье",
                trackers: [
                    Tracker(
                        id: UUID(),
                        title: "Пить воду",
                        color: UIColor(red: 0.31, green: 0.53, blue: 0.78, alpha: 1.0),
                        emoji: "💧",
                        schedule: [.monday, .tuesday, .wednesday, .thursday, .friday, .saturday, .sunday],
                        isPinned: false,
                        categoryId: categoryId1
                    ),
                    Tracker(
                        id: UUID(),
                        title: "Утренняя зарядка",
                        color: UIColor(red: 1.0, green: 0.42, blue: 0.21, alpha: 1.0),
                        emoji: "💪",
                        schedule: [.monday, .wednesday, .friday],
                        isPinned: true,
                        categoryId: categoryId1
                    )
                ]
            ),
            TrackerCategory(
                id: categoryId2,
                title: "Работа",
                trackers: [
                    Tracker(
                        id: UUID(),
                        title: "Проверить почту",
                        color: UIColor(red: 0.44, green: 0.63, blue: 0.32, alpha: 1.0),
                        emoji: "✉️",
                        schedule: [.monday, .tuesday, .wednesday, .thursday, .friday],
                        isPinned: false,
                        categoryId: categoryId2
                    )
                ]
            )
        ]
    }
}

// MARK: - Mock

final class MockTrackersViewModel: TrackersViewModelProtocol {
    var currentDate = Date()
    var searchText = ""
    var currentFilter: TrackersViewModel.FilterType = .all
    var stubVisibleCategories: [TrackerCategory] = []
    
    var visibleCategories: [TrackerCategory] { stubVisibleCategories }
    var hasTrackersOnSelectedDate: Bool { !stubVisibleCategories.isEmpty }
    var isFilterActive: Bool { false }
    
    var onFilterChanged: (() -> Void)?
    var onDataUpdate: (() -> Void)?
    
    func loadData() {}
    func updateVisibleCategories() { onDataUpdate?() }
    func isTrackerCompleted(_ trackerID: UUID, for date: Date) -> Bool { false }
    func getTotalCompletionCount(for trackerID: UUID) -> Int { 0 }
    func getCategoryTitle(for categoryId: UUID) -> String? { nil }
    func toggleTrackerCompletion(_ tracker: Tracker, for date: Date) {}
    func updateSearchText(_ text: String) {}
    func updateCurrentDate(_ date: Date) {}
    func togglePinStatus(for tracker: Tracker) {}
    func deleteTracker(_ tracker: Tracker, completion: @escaping (Result<Void, Error>) -> Void) {}
}

```

IgorChumikov commented on TrackerTests/TrackersViewControllerSnapshotTests.swift:
```diff
@@ -0,0 +1,103 @@
+    }
+}
+
+class MockTrackersViewModel: TrackersViewModelProtocol {
```
💡 Рекомендация по улучшению структуры моков

Сейчас MockTrackersViewModel находится прямо в тестовом файле.
Лучше вынести его в отдельный файл, чтобы:

📦 Разгрузить сами тесты — код мока не будет мешать читать логику теста.
🔄 Повторно использовать моки в других тестах без дублирования кода.
🧾 Соблюдать единый стиль — моки в одном месте, тесты в другом.

Предлагаемый вариант
Создать файл TrackersViewModelMock.swift в папке TrackerTests/Mocks/.
Переименовать класс в TrackersViewModelMock (Mock в конце — общепринятая практика в iOS).

IgorChumikov commented on TrackerTests/TrackersViewControllerSnapshotTests.swift:
```diff
@@ -0,0 +1,103 @@
+        assertSnapshot(of: vc, as: .image(traits: .init(userInterfaceStyle: .dark)), named: "dark_mode")
+    }
+    
+    private func setupMockData(for mockVM: MockTrackersViewModel) {
```
setupMockData лучше вынести из теста — чтобы тесты были чище и фикстуру можно было переиспользовать. главное: вне тестов нельзя дергать XCTFail, поэтому делаем либо ! на заведомо валидных UUID, либо возвращаем Result.

ниже самый удобный вариант — экстеншен на мок + фабрика. положи это в отдельный файл, например:

```

// TrackerTests/Fixtures/TrackersViewModelMock+Fixtures.swift
import UIKit
@testable import Tracker

extension TrackersViewModelMock {

    /// Заполняет мок предсказуемыми данными для снапшотов.
    /// Возвращает self, чтобы можно было чейнить/использовать в setUp.
    @discardableResult
    func applySnapshotFixture() -> Self {
        // Эти строки валидны, можно безопасно форс-анврапать
        let categoryId1 = UUID(uuidString: "00000000-0000-0000-0000-000000000001")!
        let categoryId2 = UUID(uuidString: "00000000-0000-0000-0000-000000000002")!

        stubVisibleCategories = [
            TrackerCategory(
                id: categoryId1,
                title: "Здоровье",
                trackers: [
                    Tracker(
                        id: UUID(),
                        title: "Пить воду",
                        color: UIColor(red: 0.31, green: 0.53, blue: 0.78, alpha: 1.0),
                        emoji: "💧",
                        schedule: [.monday, .tuesday, .wednesday, .thursday, .friday, .saturday, .sunday],
                        isPinned: false,
                        categoryId: categoryId1
                    ),
                    Tracker(
                        id: UUID(),
                        title: "Утренняя зарядка",
                        color: UIColor(red: 1.0, green: 0.42, blue: 0.21, alpha: 1.0),
                        emoji: "💪",
                        schedule: [.monday, .wednesday, .friday],
                        isPinned: true,
                        categoryId: categoryId1
                    )
                ]
            ),
            TrackerCategory(
                id: categoryId2,
                title: "Работа",
                trackers: [
                    Tracker(
                        id: UUID(),
                        title: "Проверить почту",
                        color: UIColor(red: 0.44, green: 0.63, blue: 0.32, alpha: 1.0),
                        emoji: "✉️",
                        schedule: [.monday, .tuesday, .wednesday, .thursday, .friday],
                        isPinned: false,
                        categoryId: categoryId2
                    )
                ]
            )
        ]

        return self
    }

    /// Удобная фабрика: сразу отдаёт мок с готовой фикстурой.
    static func snapshotFixture() -> TrackersViewModelMock {
        TrackersViewModelMock().applySnapshotFixture()
    }
}

```

как это использовать в тесте:
```
final class TrackersViewControllerSnapshotTests: XCTestCase {
    private var vm: TrackersViewModelMock!
    private var sut: TrackersViewController!

    override func setUp() {
        super.setUp()
        vm = .snapshotFixture()                 // <-- без дублирования и XCTFail
        sut = TrackersViewController(viewModel: vm)
    }

    override func tearDown() {
        sut = nil
        vm = nil
        super.tearDown()
    }

    func testLight() {
        sut.overrideUserInterfaceStyle = .light
        assertSnapshot(of: sut, as: .image(traits: .init(userInterfaceStyle: .light)))
    }

    func testDark() {
        sut.overrideUserInterfaceStyle = .dark
        assertSnapshot(of: sut, as: .image(traits: .init(userInterfaceStyle: .dark)))
    }
}
```

почему так лучше:

тест не содержит «шум» с подготовкой данных;
фикстура переиспользуемая и детерминированная;
нет зависимости от XCTest внутри фикстуры (никаких XCTFail);
единая точка изменения данных снапшота.

SergeySKRV committed:
✨  feat: Implemented review recommendations

- Improved analytics implementation using dedicated AnalyticsService with enums for events, screens and items
- Replaced all AppMetrica.reportEvent calls with AnalyticsService.shared.reportEvent maintaining type safety
- Fixed AnalyticsService.swift errors (unique enum values, added missing cases)
- Fixed tracker update logic when selecting "Trackers for today" filter
- Updated all project files with new analytics requirements
- Refactored tests using mocks and fixtures for snapshots
